blog/2021-06-14-stryker4s-sbt-performance.mdx

```diff
@@ -0,0 +1,151 @@
+.scalatest.tools.Framework@2b9d52d5)
+...
+sbt:stryker4s-core> show testGrouping
+[info] * (<default>,Vector(Test stryker4s.config.TestFilterTest : subclass(false, org.scalatest.Suite)
+, Test stryker4s.extension.FileExtensionsTest : subclass(false, org.scalatest.Suite), Test stryker4s.m
+utants.MutatorTest : subclass(false, org.scalatest.Suite), Test stryker4s.mutants.AddAllMutationsTest
+: subclass(false, org.scalatest.Suite), ...
+```
+
+We now know what tests are in your project. Let's have a look at running those tests:
+
+### 🧪 Meet: test interfaces
+
+This is something our friends in the NodeJS world are (rightfully) a little [jealous of](https://stryker-mutator.io/docs/stryker-js/plugins#test-runners). When you call `sbt test`, sbt will look for any registered test frameworks that implement the [test-interface](https://github.com/sbt/test-interface) (from `loadedTestFrameworks`). This is a small collection of interfaces that let a test framework tell sbt how to run tests in that framework. Because all test frameworks in sbt implement this interface, we have a single programmatic way to call any test framework. It doesn't matter if you're using ScalaTest, MUnit, Weaver-test or anything else. Using a bit of reflection, we can create an instance of the test framework runner and run tests:
+
+```scala
+val framework = classLoader.loadClass(testFrameworkClass).getConstructor().newInstance().asInstanceOf[Framework]
+val runner = framework.runner(args, remoteArgs, classLoader)
+
+val testTasks: Array[Task] = runner.tasks(testGrouping.taskDefs)
+
+testTasks.foreach(testTask => {
+  testTask.execute()
+})
+```
+
+### 🔀 Meet: concurrency
+
+When running your unit tests, Stryker4s activates each mutant separately and runs your unit tests against _only_ the change that mutant creates. It won't activate two mutants at the same time because there's no way to know which of the two mutants gave a result. However, there's no rule that says we can't start multiple Java processes at the same time and run those in parallel. Each process has its own separate memory, threads, and active mutant. This way, if one mutant result in an `OutOfMemoryError` it won't impact the others. Starting multiple testrunner workers also improves performance by making better use of multi-core CPUs.
+
+### 🧰 Putting it all together
+
+Stryker4s can start a new process with all the information collected above. Because Stryker4s now fully controls how it runs your tests, it can eliminate any overhead of starting new processes or looking for source code changes. It looks a little something like this:
+
+1. Start new worker processes in parallel with the full test classpath and send them all the needed test setup.
+1. Create a new instance of the test framework in the created process.
+1. For each mutant, programmatically run the tests on the worker process.
+
+<Mermaid
+  _key="sbt-new"
+  chart={`
+sequenceDiagram
```

Comment:
Suggestion:

```
sequenceDiagram
  participant Stryker4s
  participant sbt
  participant testrunner worker
  Stryker4s->>+sbt: compile and retrieve build info
  sbt-->>-Stryker4s: return project info
  Stryker4s->>+testrunner worker: start <n> new workers
  loop For each mutant
    Stryker4s->>testrunner worker: run tests
    testrunner worker-->>Stryker4s: report test result
  end
  Stryker4s->>testrunner worker: close test worker
  testrunner worker-->>-Stryker4s: 
```

Reply:
I'm confused now. In this diagram the suggestion is to have Stryker4s first and then sbt, but in the first one to have sbt first and then Stryker4s?

Reply:
Both of them should have Stryker4s first, but I didn't try them locally, just in the mermaidjs playground